File: Projects/codechef/COINS.java
package codechef;

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

class COINS {
    private static final Map<Integer, Long> dp = new HashMap<>();

    public static long exchange(int n) {
        if (n < 12) {
            return n;
        } else if (n == 12) {
            return 13;
        } else if (dp.containsKey(n)) {
            return dp.get(n);
        }
        dp.put(n, Math.max(n, exchange(n / 2) + exchange(n / 3) + exchange(n / 4)));
        return dp.get(n);
    }

    public static void main(String[] args) throws java.lang.Exception {
        Scanner stdin = new Scanner(System.in);
        while (stdin.hasNextLine()) {
            int n = Integer.parseInt(stdin.nextLine());
            System.out.println(exchange(n));
        }
    }
}

File: Projects/codechef/COINS.kt
package codechef

import java.util.*
import kotlin.math.max

object COINS {

    private val map = mutableMapOf(0 to 0L)

    private fun exchange(n: Int): Long = when (n) {
        in map -> map[n]!!
        else -> {
            val subtotal2 = exchange(n / 2)
            val subtotal3 = exchange(n / 3)
            val subtotal4 = exchange(n / 4)
            val total = subtotal2 + subtotal3 + subtotal4
            map[n] = max(n.toLong(), total)
            map[n]!!
        }
    }

    fun main(args: Array<String>) {
        val stdin = Scanner(System.`in`)
        while (stdin.hasNextLine()) {
            val n = stdin.nextLine().toInt()
            println(exchange(n))
        }
        stdin.close()
    }
}